tools/ci/templates/jobs-unity-package.yaml

```diff
@@ -46,92 +46,158 @@ jobs:
+      inPathRoot: '$(Build.SourcesDirectory)/libs/unity/library'
+      outPathRoot: '$(Build.SourcesDirectory)/libs/unity/library'
     env:
-      NPM_PUBLISH_REGISTER: $(NPM_PUBLISH_REGISTER)
-      NPM_PUBLISH_AUTH: $(NPM_PUBLISH_AUTH)
-      NPM_PUBLISH_EMAIL: $(NPM_PUBLISH_EMAIL)
-    timeoutInMinutes: 5
+      SYSTEM_ACCESSTOKEN: $(System.AccessToken)
 
-  # Publish library
-  - script: |
-      cd libs/unity/library
-      npm publish --registry $(NPM_PUBLISH_REGISTER)
-    displayName: 'Publish NPM (library)'
+  # For debugging, list the packages content after signing
+  - powershell: |
+     foreach ($f in $(Get-ChildItem -Path libs/unity -Recurse)) {
+       Write-Host $f.FullName
+     }
+    displayName: 'List packages content'
 
-  # Delete .npmrc
+  # Pack the NPM package (library)
   - pwsh: |
-      Remove-Item -Path '$(Build.SourcesDirectory)/libs/unity/library/.npmrc' -Force -ErrorAction Ignore
-    displayName: 'Delete .npmrc (library)'
-    condition: always()
+      Set-Location "libs/unity/library"
+      Invoke-Expression "npm pack ." | Tee-Object -Variable PackageName | Out-Null
+      Write-Host "Produced package '$PackageName'"
+      Write-Host "##vso[task.setvariable variable=PackedLibraryFilename]$PackageName"
+    displayName: 'Pack NPM (library)'
 
-  # Write .npmrc
-  - task: PowerShell@2
-    displayName: 'Write .npmrc (samples)'
+  # Pack the NPM package (samples)
+  - pwsh: |
+      Set-Location "libs/unity/samples"
+      Invoke-Expression "npm pack ." | Tee-Object -Variable PackageName | Out-Null
+      Write-Host "Produced package '$PackageName'"
+      Write-Host "##vso[task.setvariable variable=PackedSamplesFilename]$PackageName"
+    displayName: 'Pack NPM (samples)'
+
+  # Publish the UPM package (library)
+  - task: PublishPipelineArtifact@1
+    displayName: 'Publish artifact (library)'
     inputs:
-      targetType: 'filePath'
-      filePath: '$(Build.SourcesDirectory)/tools/ci/writeNpmrc.ps1'
-      arguments: '-Target "libs/unity/samples/.npmrc"'
-    env:
-      NPM_PUBLISH_REGISTER: $(NPM_PUBLISH_REGISTER)
-      NPM_PUBLISH_AUTH: $(NPM_PUBLISH_AUTH)
-      NPM_PUBLISH_EMAIL: $(NPM_PUBLISH_EMAIL)
-    timeoutInMinutes: 5
+      path: '$(Build.SourcesDirectory)/libs/unity/library/$(PackedLibraryFilename)'
+      artifact: 'upm_library'
 
-  # Publish samples
-  - script: |
-      cd libs/unity/samples
-      npm publish --registry $(NPM_PUBLISH_REGISTER)
-    displayName: 'Publish NPM (samples)'
+  # Publish the UPM package (samples)
+  - task: PublishPipelineArtifact@1
+    displayName: 'Publish artifact (samples)'
+    inputs:
+      path: '$(Build.SourcesDirectory)/libs/unity/samples/$(PackedSamplesFilename)'
+      artifact: 'upm_samples'
 
-  # Delete .npmrc
-  - pwsh: |
-      Remove-Item -Path '$(Build.SourcesDirectory)/libs/unity/samples/.npmrc' -Force -ErrorAction Ignore
-    displayName: 'Delete .npmrc (samples)'
-    condition: always()
+  # # Write .npmrc
```

Comment:
What's the deal with the commented code?

Reply:
That's for directly pushing to a NPM register. It was difficult to figure out some proper auth. On the other hand we might want to not push directly from the pipeline because currently YAML build pipelines don't support manual gates (manual approval button) and it's safer to be able to inspect the build and eye it one time before actually publishing definitely to a public repository. In short: that's a memo for if/when ADO supports manual gating, and in the meantime we'll publish manually for safety with the code above it.

Reply:
Add a comment explaining this (or delete)

Reply:
Fixed.